File: typescript/sdk/src/xml.ts
/**
 * XML generation for the client-side system-prompt skills catalog.
 */

import type { SkillSummary } from "./types.js";

/**
 * Escape XML special characters.
 */
export function escapeXml(text: string): string {
  return text
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&apos;");
}

/**
 * Options for generateSkillsXMLFromSummaries().
 */
export interface SkillsXmlOptions {
  /** Server name to inject when `serverInEntries` is true. */
  serverName?: string;
  /**
   * Inject `<server>{serverName}</server>` into each `<skill>` entry.
   * Default: false. The host SKILL.md flags this as a way to keep
   * first-call activation reliability ~90% for `(server, uri)` reader
   * tools — but it's not in SEP-2640, so the SDK leaves it off by default.
   */
  serverInEntries?: boolean;
}

/**
 * Generate <available_skills> XML from a client-side SkillSummary array.
 */
export function generateSkillsXMLFromSummaries(
  skills: SkillSummary[],
  options?: SkillsXmlOptions,
): string {
  const serverName = options?.serverName;
  const inEntries = options?.serverInEntries === true;
  const lines: string[] = ["<available_skills>"];

  for (const skill of skills) {
    lines.push("  <skill>");
    lines.push(`    <name>${escapeXml(skill.name)}</name>`);
    lines.push(`    <path>${escapeXml(skill.skillPath)}</path>`);
    if (serverName && inEntries) {
      lines.push(`    <server>${escapeXml(serverName)}</server>`);
    }
    if (skill.description) {
      lines.push(
        `    <description>${escapeXml(skill.description)}</description>`,
      );
    }
    lines.push(`    <uri>${escapeXml(skill.uri)}</uri>`);
    lines.push("  </skill>");
  }

  lines.push("</available_skills>");
  return lines.join("\n");
}

File: python/sdk/src/mcp_experimental_ext_skills/_xml.py
"""XML generation for the client-side system-prompt skills catalog."""

from __future__ import annotations

from ._types import SkillSummary


def _escape_xml(text: str) -> str:
    """Escape XML special characters."""
    return (
        text.replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace('"', "&quot;")
        .replace("'", "&apos;")
    )


def generate_skills_xml_from_summaries(
    skills: list[SkillSummary],
    server_name: str | None = None,
) -> str:
    """Generate ``<available_skills>`` XML from a SkillSummary list.

    When ``server_name`` is provided, each entry includes a ``<server>``
    tag so the model has the server name in context next to the URI it
    would pass to a ``(server, uri)`` reader tool.
    """
    lines: list[str] = ["<available_skills>"]
    for skill in skills:
        lines.append("  <skill>")
        lines.append(f"    <name>{_escape_xml(skill.name)}</name>")
        lines.append(f"    <path>{_escape_xml(skill.skill_path)}</path>")
        if server_name:
            lines.append(f"    <server>{_escape_xml(server_name)}</server>")
        if skill.description:
            lines.append(
                f"    <description>{_escape_xml(skill.description)}</description>"
            )
        lines.append(f"    <uri>{_escape_xml(skill.uri)}</uri>")
        lines.append("  </skill>")
    lines.append("</available_skills>")
    return "\n".join(lines)


__all__ = ["generate_skills_xml_from_summaries"]
